### PR TITLE
tree: Add 2 new public functions to lookup existing controllers

### PR DESCRIPTION
**Pull request by martin-belanger**
There is duplicate code between **libnvme** and **nvme-cli**. When trying to instantiate a controller, both libnvme and nvme-cli scan the sysfs looking for an existing controller that can be reused.

This patch adds `nvme_ctrl_find()` and `nvme_ctrl_config_match()`. These functions can be used by nvme-cli to lookup existing controllers. This will allow removing the duplicate code in nvme-cli (separate commit to follow on the nvme-cli GitHub repo).

Note that some of the existing code had to be reorganized to make it possible for the new functions to reuse that existing code.

### CODE REVIEW

File: src/nvme/private.h [hunk 1]
```diff
@@ -180,7 +180,8 @@ int json_dump_tree(nvme_root_t r);
 nvme_ctrl_t __nvme_lookup_ctrl(nvme_subsystem_t s, const char *transport,
 			       const char *traddr, const char *host_traddr,
 			       const char *host_iface, const char *trsvcid,
-			       nvme_ctrl_t p);
+			       nvme_ctrl_t p, const char *subsysnqn,
+			       bool dc_only);
```
Comment (igaw) [inline src/nvme/private.h:184]: `dc_only` could be an enum type which makes it a bit simpler to read on the caller side IMO.

Comment (martin-belanger) [reply src/nvme/private.h:184]: That's an interesting idea. I didn't really like this parameter in the first place, but we're a bit stuck. The issue is that for the longest time DCs only had the well-known NQN (nqn.2014-08.org.nvmexpress.discovery). This meant that by looking at the subsysnqn we could easily determine whether we were dealing with an I/O or Discovery controller.

With TP8013, however, DCs can now have a unique NQN. On top of that, one can create a connection to a DC using the well-known NQN, but during the identify phase, the NQN reported through the sysfs can change to a unique NQN. So the config subsysnqn may not match the NQN reported by the sysfs. There are other attributes that can be read from the sysfs to make differentiate a discovery from an I/O controller.

Long story short, we could indeed use an enum here. I can make the change.

Comment (martin-belanger) [reply src/nvme/private.h]: Actually, this can be simplified. We don't need an additional parameter to identify the controller type. I went down this path because of how it was done in nvme-cli, but the reality is that we only need to use special handling when dealing with the well-known NQN (`"nqn.2014-08.org.nvmexpress.discovery"`). For all other cases, the unique `subsysnqn` is all we need to worry about. 

Let me clean this up and resubmit.

File: src/nvme/private.h [hunk 2]
```diff
@@ -180,7 +180,7 @@ int json_dump_tree(nvme_root_t r);
 nvme_ctrl_t __nvme_lookup_ctrl(nvme_subsystem_t s, const char *transport,
 			       const char *traddr, const char *host_traddr,
 			       const char *host_iface, const char *trsvcid,
-			       nvme_ctrl_t p);
+			       nvme_ctrl_t p, const char *subsysnqn);
```
Comment (igaw) [inline src/nvme/private.h:183]: Isn't the `p` argument used as starting point of iterator? In this case I would swap the order of `p` and `subsysnqn` so that `p` stays at the last position.

Comment (martin-belanger) [reply src/nvme/private.h]: Done

**Comment by igaw** [comment]
Looks good. Thanks!
